src/OrleansRuntime/Scheduler/ShedulerExtensions.cs

```diff
@@ -7,64 +7,23 @@ internal static class SchedulerExtensions
     {
         internal static Task<T> QueueTask<T>(this OrleansTaskScheduler scheduler, Func<Task<T>> taskFunc, ISchedulingContext targetContext)
         {
-            var resolver = new TaskCompletionSource<T>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        T result = await taskFunc();
-                        resolver.TrySetResult(result);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-
-            // it appears that it's not important that we fire-and-forget asyncFunc() because we wait on the
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore()), targetContext);
-            return resolver.Task;
+            var workItem = new AsyncClosureWorkItem<T>(taskFunc);
+            scheduler.QueueWorkItem(workItem, targetContext);
+            return workItem.Task;
         }
 
         internal static Task QueueTask(this OrleansTaskScheduler scheduler, Func<Task> taskFunc, ISchedulingContext targetContext)
         {
-            var resolver = new TaskCompletionSource<bool>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        await taskFunc();
-                        resolver.TrySetResult(true);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore()), targetContext);
-            return resolver.Task;
+            var workItem = new AsyncClosureWorkItem(taskFunc);
+            scheduler.QueueWorkItem(workItem, targetContext);
+            return workItem.Task;
         }
 
         internal static Task QueueNamedTask(this OrleansTaskScheduler scheduler, Func<Task> taskFunc, ISchedulingContext targetContext, string activityName = null)
         {
-            var resolver = new TaskCompletionSource<bool>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        await taskFunc();
-                        resolver.TrySetResult(true);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore(), () => activityName), targetContext);
```

Comment:
There was a very good reason it was done this way (with Func<string>)
The string was different per activation, and we ended up with a huge number of allocated big strings, and it stressed GC. I even remember @rrector  reporting this. He spend a lot of time chasing the GC thrashing problem, some 4 years ago, maybe.
The code has changed since then obviously, so it is possible that the passed string is not activation dependent any more. But I would be much more careful just removing this.

At least, at least I suggest breaking this PR into multiple smaller ones.